### DISCUSSION TITLE
FastAPI Testing - Post Test AssertionError

### DISCUSSION THREAD

**Discussion by siva-medapati-by** [fastapi/fastapi#6786] [Questions] [answered]
`|app
||   main.py
||   __init__.py
||
|+---router
||   |   repository.py
||   |   __init__.py
||
|+---tests
||   |   test_repository.py
||   |   __init__.py`

**_main.py_**

`from fastapi import FastAPI 
from typing import Optional
from pydantic import BaseModel
import secrets

from .router import repository

tags_metadata = [
    {
        "name": "Github Repos",
        "description": "[Operations with GitHub repos]",
    }

]

app = FastAPI(
    openapi_tags=tags_metadata,
    docs_url = "/api/v1/docs",
    openapi_url = "/api/v1/openapi.json",
    redoc_url = "/api/v1/redocs",
    title = "ALM API",
    version = "1.0",
    description = "Provides API for Application Lifecycle Management",
)
app.include_router(
    repository.router,
    tags=["Github Repos"],
)
`
**_repository.py_**

`from fastapi import APIRouter, status, HTTPException
from fastapi.responses import JSONResponse
from dotenv import load_dotenv
from os import environ
import requests
import os
from typing import Optional

router = APIRouter()

load_dotenv()
AUTH_TOKEN="Bearer "+environ.get('GITHUB_PAT')

# create github repo
@router.post("/orgs/{org}/repos", status_code=status.HTTP_201_CREATED) 
def create_github_repo(org: str,name: str, auto_init: Optional[bool] = 'true', accept:Optional[str] = 'application/vnd.github.v3+json'):
    url = "https://api.github.com/orgs/{}/repos".format(org)
    response = requests.post(url, json={'name': name, 'auto_init': auto_init}, headers={'Authorization': AUTH_TOKEN})
    return(response.json()) `

**_test_repository.py_**

`from fastapi import FastAPI, APIRouter
from fastapi.testclient import TestClient
from dotenv import load_dotenv
from os import environ
from app.router.repository import router
from app.main import app

client = TestClient(app)

load_dotenv()
AUTH_TOKEN="Bearer "+environ.get('GITHUB_PAT')

print(client.__dict__)

def test_create_github_repo():
    response = client.post('/orgs/ALM-Demo/repos', json={'name': 'demo', 'auto_init': 'true'}, headers={'Authorization': AUTH_TOKEN})
    assert response.status_code == 200`

`(env) PS C:\Users\1024257\GH\plat-lext-alm-lifecycle> pytest
=================================================== test session starts ====================================================
platform win32 -- Python 3.9.2, pytest-6.2.3, py-1.10.0, pluggy-0.13.1
rootdir: C:\Users\1024257\GH\plat-lext-alm-lifecycle
collected 1 item

app\tests\test_repository.py F                                                                                        [100%]

========================================================= FAILURES =========================================================
_________________________________________________ test_create_github_repo __________________________________________________

    def test_create_github_repo():
        response = client.post('/orgs/ALM-Demo/repos', json={'name': 'demo', 'auto_init': 'true'}, headers={'Authorization': 
AUTH_TOKEN})
>       assert response.status_code == 200
E       assert 422 == 200
E        +  where 422 = <Response [422]>.status_code

app\tests\test_repository.py:17: AssertionError
================================================= short test summary info ================================================== 
FAILED app/tests/test_repository.py::test_create_github_repo - assert 422 == 200
==================================================== 1 failed in 1.00s ===================================================== `

**Comment by insomnes** [comment, marked as answer]
@siva-medapati-by hello!
You should use text between ``` for multiline code blocks (and python3 immediately after them for syntax highlight).
For your case
```python3
@router.post("/orgs/{org}/repos", status_code=status.HTTP_201_CREATED)
def create_github_repo(org: str,name: str, auto_init: Optional[bool] = 'true', accept:Optional[str] = 'application/vnd.github.v3+json'):
    url = "https://api.github.com/orgs/{}/repos".format(org)
    response = requests.post(url, json={'name': name, 'auto_init': auto_init}, headers={'Authorization': AUTH_TOKEN})
    return(response.json())
```
You make `name` and `auto_init` [query parameters.](https://fastapi.tiangolo.com/tutorial/query-params/)
And in test you send them as json in request body:
```python3
def test_create_github_repo():
    response = client.post('/orgs/ALM-Demo/repos', json={'name': 'demo', 'auto_init': 'true'}, headers={'Authorization': AUTH_TOKEN})
    assert response.status_code == 200
```

If you want to accept them as json in [request body](https://fastapi.tiangolo.com/tutorial/body/) at your handler function, you should set them as pydatnic model attributes.

**Comment by siva-medapati-by** [comment]
Thanks @insomnes 
I am able to run the test now by passing them as query params.

**Comment by tiangolo** [comment]
Thanks for the help here @insomnes ! 👏  🙇 

Thanks for reporting back and closing the issue @siva-medapati-by  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
